### PR TITLE
Allow segments in a light partition to be reversed

### LINKED ISSUE #1011: Add ability to reverse addressable light strips
**Describe the problem you have/What new integration you would like**

I have a few LED strips around the house I control with ESPHome. Works nicely! Some of the strips are oriented left-to-right whereas others are oriented right-to-left (simply due to where I have power available). It would be very useful to have the ability to 'reverse' the strips (when it comes to addressable LED features) so I can control them all in the same way. In particular I would like to use this with WLED.

**Please describe your use case for this integration and alternatives you've tried:**

Alternatives are: swapping the LED strip physically (requires cabling for power), controlling them differently/swapping in the controlling software (complicates things).

**Additional context**

Suggestion would be to add a `reverse: true` parameter to `light` (or, if specific to wled, there).

### PR DESCRIPTION
**Pull request by jimbauwens**
## Description:
This commit adds the 'reverse' boolean option to segments of a light partition. By specifying the reverse option, a particular segment can be addressed in reverse order: e.g. in a segment of 10 lights, index 0 becomes 9, index 9 becomes 0.

This allows for the simplified placing of led strips without needing to run extra data lines and making sure all the strips line up logically.

**Related issue (if applicable):** fixes [issue 1011 (Add ability to reverse addressable light strips)](https://github.com/esphome/feature-requests/issues/1011)

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by balrog-kun — COMMENTED** [review]
Nice work.  I had an untested patch to achieve the same effect although less explcitly, by just speifying a `from` number higher than the `to` number.

I read through your changes and they look good to me FWIW.

File: esphome/components/partition/light.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def validate_from_to(value):
         cv.Required(CONF_ID): cv.use_id(light.AddressableLightState),
         cv.Required(CONF_FROM): cv.positive_int,
         cv.Required(CONF_TO): cv.positive_int,
+        cv.Optional(CONF_REVERSE, default=False): cv.boolean
```
Comment (OttoWinter) [inline esphome/components/partition/light.py:24]: Nit: I think it would make more sense to call this config option `reversed` (note the **d**).

**Comment by slimcdk** [comment]
Why limit this to partition lights? I think it would benefit more to have it directly on the neopixelbus or fastled integrations

**Comment by jimbauwens** [comment]
> Why limit this to partition lights? I think it would benefit more to have it directly on the neopixelbus or fastled integrations

Partitions allow you to remap any addressable light, including neopixelbus and fastled integrations. It makes sense to put the option here, as it fits with the concept of virtual remapping. This way the neopixelbus and fastled integrations can remain to be a correct mapping of the hardware, and you use portions to remap them when needed.

File: esphome/components/light/effects.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 from esphome.const import CONF_NAME, CONF_LAMBDA, CONF_UPDATE_INTERVAL, CONF_TRANSITION_LENGTH, \
     CONF_COLORS, CONF_STATE, CONF_DURATION, CONF_BRIGHTNESS, CONF_RED, CONF_GREEN, CONF_BLUE, \
     CONF_WHITE, CONF_ALPHA, CONF_INTENSITY, CONF_SPEED, CONF_WIDTH, CONF_NUM_LEDS, CONF_RANDOM, \
-    CONF_SEQUENCE
+    CONF_SEQUENCE, CONF_REVERSE
```
Comment (jesserockz) [inline esphome/components/light/effects.py:7]:
```suggestion
    CONF_SEQUENCE, CONF_REVERSED
```

File: esphome/components/partition/light.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import esphome.codegen as cg
 import esphome.config_validation as cv
 from esphome.components import light
-from esphome.const import CONF_FROM, CONF_ID, CONF_SEGMENTS, CONF_TO, CONF_OUTPUT_ID
+from esphome.const import CONF_FROM, CONF_ID, CONF_SEGMENTS, CONF_TO, CONF_OUTPUT_ID, CONF_REVERSE
```
Comment (jesserockz) [inline esphome/components/partition/light.py:4]:
```suggestion
from esphome.const import CONF_FROM, CONF_ID, CONF_SEGMENTS, CONF_TO, CONF_OUTPUT_ID, CONF_REVERSED
```

File: esphome/components/partition/light.py (hunk 1 above)
Comment (jesserockz) [inline esphome/components/partition/light.py:24]:
```suggestion
        cv.Optional(CONF_REVERSED, default=False): cv.boolean
```

File: esphome/components/partition/light.py [hunk 4]
```diff
@@ -30,7 +31,7 @@ def to_code(config):
     for conf in config[CONF_SEGMENTS]:
         var = yield cg.get_variable(conf[CONF_ID])
         segments.append(AddressableSegment(var, conf[CONF_FROM],
-                                           conf[CONF_TO] - conf[CONF_FROM] + 1))
+                                           conf[CONF_TO] - conf[CONF_FROM] + 1, conf[CONF_REVERSE]))
```
Comment (jesserockz) [inline esphome/components/partition/light.py:34]:
```suggestion
                                           conf[CONF_TO] - conf[CONF_FROM] + 1, conf[CONF_REVERSED]))
```

File: esphome/const.py [hunk 5]
```diff
@@ -435,6 +435,7 @@
 CONF_RESTORE_STATE = 'restore_state'
 CONF_RESTORE_VALUE = 'restore_value'
 CONF_RETAIN = 'retain'
+CONF_REVERSE = 'reverse'
```
Comment (jesserockz) [inline esphome/const.py:438]:
```suggestion
CONF_REVERSED = 'reversed'
```

File: esphome/components/light/effects.py [hunk 6]
```diff
@@ -4,7 +4,7 @@
 from esphome.const import CONF_NAME, CONF_LAMBDA, CONF_UPDATE_INTERVAL, CONF_TRANSITION_LENGTH, \
     CONF_COLORS, CONF_STATE, CONF_DURATION, CONF_BRIGHTNESS, CONF_RED, CONF_GREEN, CONF_BLUE, \
     CONF_WHITE, CONF_ALPHA, CONF_INTENSITY, CONF_SPEED, CONF_WIDTH, CONF_NUM_LEDS, CONF_RANDOM, \
-    CONF_SEQUENCE
+    CONF_SEQUENCE, CONF_REVERSED
```
Comment (jesserockz) [inline esphome/components/light/effects.py:7]:
```suggestion
    CONF_SEQUENCE
```

File: esphome/components/light/effects.py [hunk 7]
```diff
@@ -14,7 +14,6 @@
     AutomationLightEffect, ESPColor
 
 CONF_ADD_LED_INTERVAL = 'add_led_interval'
-CONF_REVERSE = 'reverse'
 CONF_MOVE_INTERVAL = 'move_interval'
```
Comment (jesserockz) [inline esphome/components/light/effects.py:17]:
```suggestion
CONF_REVERSE = 'reverse'
CONF_MOVE_INTERVAL = 'move_interval'
```

File: esphome/const.py [hunk 8]
```diff
@@ -470,6 +470,7 @@
 CONF_RESTORE_STATE = "restore_state"
 CONF_RESTORE_VALUE = "restore_value"
 CONF_RETAIN = "retain"
+CONF_REVERSED = "reversed"
 CONF_REVERSE_ACTIVE_ENERGY = "reverse_active_energy"
```
Comment (jesserockz) [inline esphome/const.py:474]:
```suggestion
CONF_REVERSE_ACTIVE_ENERGY = "reverse_active_energy"
CONF_REVERSED = "reversed"
```

**Review by jesserockz — APPROVED** [review]

**Comment by pixelspark** [comment]
This is awesome! Thanks all.

**Pull request merged**